docs: python debugger guide [ch5148] (#478)
windmilleng/tilt.build: docs/debuggers_python.md

---
title: Python Debuggers in Tilt
layout: docs
---

> Debuggers are an invaluable part of many development workflows. When you're running your app(s) in Kubernetes, using debuggers suddenly gets much harder, but Tilt's port forwarding functionality makes it easy to use any debugger that exposes a port for external connect.

Reviewer: ```suggestion
Debuggers are an invaluable part of many development workflows, and running your app(s) in Kubernetes makes using debuggers seemingly harder. But if your debugger works remotely, Tilt's port forwarding makes it easy to connect to it.
```

Author: reworked this paragraph a bit before i saw this comment, lmk what you think

Reviewer: SGTM!

> At the current moment, our recommended Python debugger for use with Tilt is [`remote-pdb`](https://pypi.org/project/remote-pdb/) (we will list other debugger options here as we validate and write guides for them).

Author: good call, done

## Debugging your project with Tilt and a remote debugger

> I'll be referring to [this example repo](https://github.com/windmilleng/debugger-examples/tree/master/python/remote-pdb), which is configured to use `remote-pdb`. Clone it and follow along!

Reviewer: +1

> I'll be referring to [this example repo](https://github.com/windmilleng/debugger-examples/tree/master/python/remote-pdb), which is configured to use `remote-pdb`. Clone it and follow along!

Reviewer: ```suggestion
We'll use [this example repo](https://github.com/windmilleng/debugger-examples/tree/master/python/remote-pdb), which is configured to use `remote-pdb`. Clone it and follow along!
```

Author: 😅I LIKE USING LOTS OF WORDS THEY MAKE ME SOUND SMART

Reviewer: Hah!

Seriously though, this "I will be doing" instead of "I will do" construction that you see in English all the time... it sounds *so* weird to non-native speakers.

> I'll be referring to [this example repo](https://github.com/windmilleng/debugger-examples/tree/master/python/remote-pdb), which is configured to use `remote-pdb`. Clone it and follow along!

Author: ha good point, will do

1. Decide what port you'll expose the debugger on. For this example, I'm using 5555.
2. Add your debugger of choice in [requirements.txt](https://github.com/windmilleng/debugger-examples/blob/master/python/remote-pdb/requirements.txt).
3. Expose your chosen port as a `containerPort` in the [Kubernetes YAML](https://github.com/windmilleng/debugger-examples/blob/master/python/remote-pdb/kubernetes.yaml) of the Deployment or Pod that you want to debug (in addition to any ports you're already exposing for normal development; e.g. here we expose port 8000 just to see the app running):
    ```yaml
    kind: Deployment
    ...
    spec:
        spec:
          containers:
            - name: example-python
              image: example-python-image
              ports:
                - containerPort: 5555
                - containerPort: 8000
    ```
4. Tell Tilt to port-forward that port in the [Tiltfile](https://github.com/windmilleng/debugger-examples/blob/master/python/remote-pdb/Tiltfile) (again, in addition to any ports you're already forwarding for normal development):
    ```python
   k8s_resource('example-python', port_forwards=[
       '8000:8000',  # app

>        '5555:5555',  # debugger

Reviewer: I like the comments on each line that explain which port is which, but maybe `[8000, 5555]` would seem simpler, less intimidating?

Author: forgot that syntax was legal. done.

   ])
    ```

### You're ready to debug!

> Congrats, now you can add breakpoints with your chosen debugger, and connect to it at your chosen port however is appropriate. Check out [this example repo](https://github.com/windmilleng/debugger-examples/tree/master/python/remote-pdb) to try it with `remote-pdb`:

Reviewer 2: I tried this example myself, and wasn't able to get it to work. 

`socat readline` isn't supported on debian machines

telnet and nc both give me weird port-forwarding errors that look like this:
https://cloud.tilt.dev/snapshot/AZ7hnOcLCHKhvOIsTX4=

am I missing something?

Reviewer: Here, with Ubuntu, I was able to connect to the debugger (but I'm not familiar with Python tooling so I didn't try doing anything with it). But yeah if Debian is not compatible, that's an issue.

Reviewer 2: ya, fwiw, any time I see in a guide "you can use any of these 3 options!" without any guidance of which one I should choose, it trips my alarm bells that they're all probably bad/broken in some subtle way (and that's why the author didn't choose one)

Author: added a bit about tripping the debugger before you try to connect

1. Insert the breakpoint:
    ```python

>    from remote_pdb import RemotePdb

Reviewer 2: do i need to add this to my code before i start debugging? in the example repo it was commented out, and I wasn't sure if i needed to uncomment it

Author: hah yeah i fixed this
